### ISSUE TITLE
Overlapping overload check differs from mypy in some cases

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#1546]
Pyright's checks for overlapping overloads differs in some ways from mypy's. That inconsistency is causing problems for typeshed checking. Pyright should be consistent with mypy in this case.

Here are some examples:

```python
# pyright: strict

from typing import Literal, TypeVar, Union, overload
@overload
def foo1(x: int) -> int: ...
@overload
def foo1(x: float) -> int: ...

# Mypy: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader
# Pyright: Overload 2 for "foo2" will never be used because its parameters overlap overload 1
@overload
def foo2(x: float) -> int: ...
@overload
def foo2(x: int) -> int: ...

# Mypy: Overloaded function signatures 1 and 2 overlap with incompatible return types
# Pyright: Overload 1 for "foo3" overlaps overload 2 and returns an incompatible type
@overload
def foo3(x: Literal[3]) -> int: ...
@overload
def foo3(x: int) -> str: ...

# Mypy: No error
# Pyright: Overload 1 for "foo4" overlaps overload 2 and returns an incompatible type
@overload
def foo4(x: int) -> int: ...
@overload
def foo4(x: float) -> str: ...

# Mypy: Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader
# Pyright: No error
_T = TypeVar("_T")
@overload
def foo5(x: _T) -> Union[int, _T]: ...
@overload
def foo5(x: int) -> str: ...
```

**Comment by erictraut** [comment]
The spec for mypy's behavior can be found here: https://github.com/python/typing/issues/253#issuecomment-389262904

**Comment by erictraut** [comment]
I've fixed two bugs in pyright and reported one bug in mypy (https://github.com/python/mypy/issues/10143). The two bugs in pyright will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.116, which I just published. It will also be in the next release of pylance.
